### DISCUSSION TITLE
"is" and "as"

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#2815] [Q&A]
I have always understood the following code snippets to be equivalent:

````
//Method 1
object o = ...;
var x = o as Interface;
if(x != null){
  var y = x;
  ...
}
````

and:
````
//Method 2
object o = ...;
if(o is Interface y){ 
  ...
}
````

However, I recently ran into a situation involving COM interfaces where Method1 would yield true but Method2 would yield false.

How are ````is```` and ````as```` related, specifically as it relates to COM interfaces?

**Comment by TonyValenti** [comment]
Err.  Minor error in code snipped.  Fixed online.

**Comment by CyrusNajmabadi** [comment]
> However, I recently ran into a situation involving COM interfaces where Method1 would yield true but Method2 would yield false.

It might be interesting to see the different IL you're seeing in this case.

**Comment by HaloFour** [comment]
Unless the C# compiler is emitting different IL specifically for COM interfaces they should be roughly equivalent.  A quick&dirty example on [SharpLab](https://sharplab.io/#v2:EYLgHgbALANALiAlgGwD4AEBMBGAsAKAIPQGYACLMgMQHsayBvAsli8xAOzjIDUBDZAFcApozIBzYXADcZAL4EFhfKQqYyAYUbNWqztwCy2ABQB5YACthAY240AlNvysXZAG58ATmQBmdMgC8ZPR8AM7UdNI6rmSIPmTGfvQAhEEcgsjIjkzOMa7oAOy+dAB0/ELCUbkxSnkURQAMVS61rNFssVxkBphmljZ22e0ucQn0iOG09ElD1XX1xTRlAiLNea0xhWRN7UpyQA=) yielded very similar IL:

Method 1:
```
        IL_0000: ldarg.1
        IL_0001: isinst Foo
        IL_0006: stloc.0
        IL_0007: ldloc.0
        IL_0008: brfalse.s IL_0011

        IL_000a: ldloc.0
``` 

Method 2:
```
        IL_0000: ldarg.1
        IL_0001: isinst Foo
        IL_0006: dup
        IL_0007: stloc.0
        IL_0008: brfalse.s IL_0011

        IL_000a: ldloc.0
```

**Comment by RikkiGibson** [comment]
This looks like a compiler bug, it would be greatly appreciated if you could provide a reproducer and post it as an issue in https://github.com/dotnet/roslyn

**Comment by yaakov-h** [comment]
FWIW I can't reproduce this using `CERTENROLLLib.IAlternativeName` as the interface. I'm seeing identical IL in both scenarios.

**Comment by TonyValenti** [comment]
I'll try to put together an example later today.

On Sat, Sep 21, 2019 at 5:53 AM Yaakov <notifications@github.com> wrote:

> FWIW I can't reproduce this using CERTENROLLLib.IAlternativeName as the
> interface. I'm seeing identical IL in both scenarios.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dotnet/csharplang/issues/2815?email_source=notifications&email_token=ADH2UVQCWY4KRDIOPVMZOO3QKX4MLA5CNFSM4IY2YNF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7IPNSQ#issuecomment-533788362>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADH2UVX22FXGFNS4WQIP5LLQKX4MLANCNFSM4IY2YNFQ>
> .
>
-- 
Tony Valenti

Schedule a meeting with me
<https://www.fasterlaw.com/book-online/tony-valenti/book>
